util/histogram_windowing.h

```diff
@@ -0,0 +1,80 @@
+//  Copyright (c) 2013, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include "util/histogram.h"
+#include "rocksdb/env.h"
+
+namespace rocksdb {
+
+class HistogramWindowingImpl : public Histogram
+{
+public:
+  HistogramWindowingImpl();
+  HistogramWindowingImpl(uint64_t num_windows,
+                         uint64_t micros_per_window,
+                         uint64_t min_num_per_window);
+
+  HistogramWindowingImpl(const HistogramImpl&) = delete;
+  HistogramWindowingImpl& operator=(const HistogramImpl&) = delete;
+
+  ~HistogramWindowingImpl();
+
+  virtual void Clear() override;
+  virtual bool Empty() const override;
+  virtual void Add(uint64_t value) override;
+  virtual void Merge(const Histogram& other) override;
+  void Merge(const HistogramWindowingImpl& other);
+
+  virtual std::string ToString() const override;
+  virtual const char* Name() const override { return "HistogramWindowingImpl"; }
+  virtual uint64_t min() const override { return stats_.min(); }
+  virtual uint64_t max() const override { return stats_.max(); }
+  virtual uint64_t num() const override { return stats_.num(); }
+  virtual double Median() const override;
+  virtual double Percentile(double p) const override;
+  virtual double Average() const override;
+  virtual double StandardDeviation() const override;
+  virtual void Data(HistogramData* const data) const override;
+
+private:
+  void TimerTick();
+  void SwapHistoryBucket();
+  inline uint64_t current_window() const {
+    return current_window_.load(std::memory_order_relaxed);
+  }
+  inline uint64_t last_swap_time() const{
+    return last_swap_time_.load(std::memory_order_relaxed);
+  }
+
+  Env* env_;
+  std::mutex mutex_;
```

Comment:
Yes. I have also thought about this. 
One way to implement this is to have a queue: new data is en-queued, old data is expired by de-queue, another worker thread is used to periodical aggregate the values.
The problem is that 1) storing raw data is expensive, 10k QPS * 8 byte * 300 sec = 24 MB 
2) another aggregation thread is needed.